framework/src/base/FEProblem.C

```diff
@@ -555,13 +560,23 @@ void FEProblem::initialSetup()
       computeUserObjects(EXEC_TIMESTEP_BEGIN);
       computeUserObjects(EXEC_LINEAR);
     }
-    Moose::setup_perf_log.pop("Initial computeUserObjects()","Setup");
-  }
 
-  _nl.initialSetupBCs();
-  _nl.initialSetupKernels();
+    // Perform an initial residual evaluation to set save-in variables
+    // Note: time kernels need not to be evaluated on initial because their contribution should be just zero.
+    //       They actually cannot be evaluated because we set the time step size to zero on initial.
+    if (_nl.hasSaveIn())
+    {
+      _nl.computeResidual(*_nl.sys().rhs, Moose::KT_NONTIME);
+      // We want to call post user object evaluation one more time in case they depends on the save-in variables.
+      computeUserObjects(EXEC_INITIAL, UserObjectWarehouse::POST_AUX);
```

Comment:
@YaqiWang this makes me think that we need to rethink this whole thing.  We can't just arbitrarily decided to do this again.

Maybe someone else needs UserObjects computed in order to compute their residual... or maybe their initial user object computation is something that takes 5 hours to complete (like reading in a huge datastructure) and this would cause it to happen again.

We can't just do these again.  We've tried to remove all of this kind of stuff out of MOOSE.

This kind of makes me think that we're just going to need to find a solution more specific to your eigenvalue stuff.  Like putting a virtual function in here somewhere where you can do another residual computation and another UserObject computation in a custom executioner or something.